source/endpoints_ruleset.c

```diff
@@ -0,0 +1,935 @@
+
+    *out_reference = NULL;
+    struct aws_json_value *ref_node = aws_json_value_get_from_object(node, aws_byte_cursor_from_c_str("ref"));
+    if (ref_node != NULL) {
+        if (!aws_json_value_is_string(ref_node)) {
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        }
+
+        struct aws_byte_cursor ref_cur;
+        aws_json_value_get_string(ref_node, &ref_cur);
+        *out_reference = aws_string_new_from_cursor(allocator, &ref_cur);
+        return AWS_OP_SUCCESS;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_parse_expr(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *node,
+    struct aws_endpoints_expr *expr);
+
+static int s_on_expr_element(
+    size_t idx,
+    const struct aws_json_value *value_node,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)idx;
+    (void)out_should_continue;
+    struct array_parser_wrapper *wrapper = user_data;
+
+    struct aws_endpoints_expr expr;
+    AWS_ZERO_STRUCT(expr);
+    if (s_parse_expr(wrapper->allocator, value_node, &expr)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse expr.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    aws_array_list_push_back(wrapper->array, &expr);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_parse_expr(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *node,
+    struct aws_endpoints_expr *expr) {
+    /* TODO: this recurses. in practical circumstances depth will never be high,
+    but we should still consider doing iterative approach */
+    if (aws_json_value_is_string(node)) {
+        struct aws_byte_cursor cur;
+        aws_json_value_get_string(node, &cur);
+        expr->type = AWS_ENDPOINTS_EXPR_STRING;
+        expr->e.string = aws_string_new_from_cursor(allocator, &cur);
+        return AWS_OP_SUCCESS;
+    } else if (aws_json_value_is_number(node)) {
+        double number;
+        aws_json_value_get_number(node, &number);
+        expr->type = AWS_ENDPOINTS_EXPR_NUMBER;
+        expr->e.number = number;
+        return AWS_OP_SUCCESS;
+    } else if (aws_json_value_is_boolean(node)) {
+        bool v;
+        aws_json_value_get_boolean(node, &v);
+        expr->type = AWS_ENDPOINTS_EXPR_BOOLEAN;
+        expr->e.boolean = v;
+        return AWS_OP_SUCCESS;
+    } else if (aws_json_value_is_array(node)) {
+        expr->type = AWS_ENDPOINTS_EXPR_ARRAY;
+        size_t num_elements = aws_json_get_array_size(node);
+        aws_array_list_init_dynamic(&expr->e.array, allocator, num_elements, sizeof(struct aws_endpoints_expr));
+        if (s_init_array_from_json(allocator, node, &expr->e.array, s_on_expr_element)) {
+            aws_endpoints_expr_cleanup(expr);
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse array value type.");
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
```

Comment:
I'm worried that ALSO we might need to clean up this array if something else goes wrong later in the function

maybe just `goto error` here, and always call `aws_endpoints_expr_cleanup(expr);` from there. That's my general approach: always try to clean up everything

like, what if we parse a list like `"ref": ["i", "am", "not", "a", "ref"]`? We'd succesfully parse the list, but then the call to `s_try_parse_reference()` would fail and we'd leak this memory

Reply:
Parsing array is terminal here and it returns either on success or failure, so failing to parse ref should not leak.
This function already has an on_error, so moved cleanup there.